llvm/lib/CodeGen/PeepholeOptimizer.cpp

```diff
@@ -2012,6 +2008,17 @@ ValueTrackerResult ValueTracker::getNextSourceFromRegSequence() {
 
     unsigned ComposedDefInSrcReg1 =
         TRI->composeSubRegIndices(RegSeqInput.SubReg, ReverseDefCompose);
+
+    // TODO: We should be able to defer checking if the result register class
+    // supports the index to continue looking for a rewritable source.
+    //
+    // TODO: Should we modify the register class to support the index?
+    const TargetRegisterClass *SrcRC = MRI.getRegClass(RegSeqInput.Reg);
+    const TargetRegisterClass *SrcWithSubRC =
+        TRI->getSubClassWithSubReg(SrcRC, ComposedDefInSrcReg1);
```

Comment:
What is the return in the test case given?
Is it `null` -- is the comparison with the SrcRC important?

Reply:
Yes, the class matters:

Here it's SrcWithSubRC: DPR_VFP2 SrcRC: DPR. We would need to restrict the register class to the subclass with the subreg